### PR TITLE
ISPN-15143 Utilise Junit5 TestNG engine to run tests

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-15143

The CDI integration tests still utilise vanilla TestNG as it's necessary to convert the tests to utilise `org.jboss.arquillian.junit5` in order for them to work with JUnit 5 as expected.

@tristantarrant  I haven't applied `rerunFailingTestsCount` yet. What value do we think is acceptable, 2/3? I think higher than that and we're potentially ignoring tests that are definitely too flaky to be acceptabble.

### CODE REVIEW

**Comment by ryanemerson** [comment]
It seems the total number of tests being executed (or at least reported) is much less than prior to the migration. Investigating...

**Comment by tristantarrant** [comment]
I'd set `rerunFailingTestsCount`  to 2

**Comment by ryanemerson** [comment]
> It seems the total number of tests being executed (or at least reported) is much less than prior to the migration. Investigating...

The `org.infinispan.commons.test.TestNGTestListener` weren't being attached correctly, so the total number of tests being executed was massively under reported.

**Comment by ryanemerson** [comment]
@tristantarrant I think we're good to go. I have set `rerunFailingTestsCount=2` for both surefire and failsafe tests, this means that a failing test will be attempted up to 3 times.

To solve the issue of Flaky tests being hidden in the Jenkins UI I have utilised the [flaky-test-extractor-maven-plugin](https://github.com/zeebe-io/flaky-test-extractor-maven-plugin). Consequently builds with flaky tests still won't report :green_circle:, however we can distinguish between hard and flaky failures in the Jenkins UI as the latter are marked with `(Flaky Test)`, see screenshot below:

Screenshot from 2023-09-15 09-15-15

**Comment by ryanemerson** [comment]
I'm not sure the history of the files, but maybe we want to remove the `Polarion` suffix from the XML classes ... I don't think this is relevant anymore :man_shrugging:

**Comment by alanfx** [comment]
Hey @ryanemerson, can you tell me more about the XML files with a `Polarion` suffix? Polaroid is still in use, unfortunately

**Comment by ryanemerson** [comment]
@alanfx I'm referring to the `PolarionJUnitXMLReporter` and `PolarionJUnitXMLWriter`. Looking at the git history it seems you contributed this in 2017. What is `Polarion`/`Polaroid`?

The changes I have made result in a JUnit xml file being created per Test class instead of a single XML per module, as well as updating the writer to handle flaky test results.

**Comment by alanfx** [comment]
I'm glad you are completely unaware of what Polarion is! (My uncle worked at Polaroid, so I know what this is too!) Polarion is the test case management tool used by Red Hat. DG basically writes our Jenkins job results there, since our testing is mostly automated. I'll have to go back myself and remember why `PolarionJUnitXMLReporter` and `PolarionJUnitXMLWriter` were required though.

**Comment by ryanemerson** [comment]
> I think these classes were needed to produce Unit XML files for TestNG tests. With this PR will the TestNG tests produce those files? If not, then these TestNG listeners are still required.

Yes we still produce JUnit xml for TestNG tests, however we're now producing a .xml file per class not per module. If a per module file is required I can make that behaviour configurable for downstream. @Crumby Any insights/thoughts?

**Comment by alanfx** [comment]
As long as the XML files are being generated, it doesn't matter if it is one big file or a bunch of little ones!

**Comment by pruivo** [comment]
The number of tests does not match and I can't tell if it is TestNG reporting too many tests.

Looking at a single test,  in PR [AsymmetricRoutingDefaultLocalCacheTest](https://ci.infinispan.org/job/Infinispan/job/PR-11253/38/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/) vs main [AsymmetricRoutingDefaultLocalCacheTest](https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/)

This PR reports ~30k -> https://ci.infinispan.org/job/Infinispan/job/PR-11253/38/testReport/
Main branch report ~54k -> https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/BaseMultiServerRemoteIteratorTest.java [hunk 1]
```diff
@@ -70,7 +80,7 @@ public void testBatchSizes() {
       int maximumBatchSize = 120;
       RemoteCache<Integer, AccountHS> cache = clients.get(0).getCache();
 
-      populateCache(CACHE_SIZE, this::newAccount, cache);
+      populateCache(CACHE_SIZE, org.infinispan.client.hotrod.impl.iteration.Util::newAccount, cache);
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/BaseMultiServerRemoteIteratorTest.java:83]: nitpick
```suggestion
      populateCache(CACHE_SIZE, Util::newAccount, cache);
```

same for below

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+   void add(ITestResult tr) {
+      switch (tr.getStatus()) {
+         case ITestResult.SUCCESS:
+            status = failures.isEmpty() ? Status.SUCCESS : Status.FLAKY;
+            successes.incrementAndGet();
+            break;
+         case ITestResult.FAILURE:
+            failures.add(tr.getThrowable());
+            status = Status.FAILURE;
```
Comment (pruivo) [inline commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java:43]: shouldn't be `FLAKY` if `successes > 0`?

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.testng.ITestResult;
+
+public class PolarionJUnitTest {
+
+   public enum Status {SUCCESS, FLAKY, FAILURE, ERROR, SKIPPED}
+
+   final String name;
+   final String clazz;
+   final List<Throwable> failures = new ArrayList<>();
```
Comment (pruivo) [inline commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java:16]: The `AtomicInteger/Long` suggests that this class may be updated concurrently. Should we replace `ArrayList` with `CopyOnWriteArrayList`?

**Review by pruivo — COMMENTED** [review]
LGTM to me, but left some suggestions. Looking forward to having it merged!

**Comment by ryanemerson** [comment]
> The number of tests does not match and I can't tell if it is TestNG reporting too many tests.
> 
> Looking at a single test, in PR [AsymmetricRoutingDefaultLocalCacheTest](https://ci.infinispan.org/job/Infinispan/job/PR-11253/38/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/) vs main [AsymmetricRoutingDefaultLocalCacheTest](https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/)
> 
> This PR reports ~30k -> https://ci.infinispan.org/job/Infinispan/job/PR-11253/38/testReport/ Main branch report ~54k -> https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/

The discrepancy is due to `main` reporting TestNG tests twice. Using the AsymmetricRoutingDefaultLocalCacheTest example, we can see on [main](https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/) that we have two entries for the same test [1](https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/client_hotrod_AsymmetricRoutingDefaultLocalCacheTest/), [2](https://ci.infinispan.org/job/Infinispan/job/main/lastCompletedBuild/testReport/org.infinispan.client.hotrod/AsymmetricRoutingDefaultLocalCacheTest/testRequestRouting/).

See my earlier comment: https://github.com/infinispan/infinispan/pull/11253#issuecomment-1720874889.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/BaseMultiServerRemoteIteratorTest.java (hunk 1 above)
Comment (ryanemerson) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/BaseMultiServerRemoteIteratorTest.java:83]: We can't do this as we're already importing `org.infinispan.commons.util.Util`.

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java (hunk 2 above)
Comment (ryanemerson) [reply commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java:43]: Flaky tests are executed until they pass, or the max number of attempts has been reached. Therefore for a test to be FLAKY, the last encountered test must be a success.

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java (hunk 3 above)
Comment (ryanemerson) [reply commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java:16]: Good point. As we're frequently writing, wouldn't using `Collections.synchronizeList()` be more efficient as I can't imagine contention will be an issue.

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java [hunk 4]
```diff
@@ -347,21 +229,21 @@ private void checkDuplicatesAndAdd(ITestResult tr) {
       // Need fully qualified name to guarantee uniqueness in the results map
       String instanceName = tr.getInstanceName();
       String key = instanceName + "." + testName(tr);
+      PolarionJUnitTest meta;
       if (m_allTests.containsKey(key)) {
-         if (tr.getMethod().getCurrentInvocationCount() == 1 && tr.isSuccess()) {
+         meta = m_allTests.get(key);
+         if (rerunFailingTestsCount == 0) {
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java:235]: I think that should work :slightly_smiling_face:

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java (hunk 3 above)
Comment (pruivo) [reply commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitTest.java:16]: I'm not concerned about perf here. I'm ok using `Collections.synchronizeList` here too.

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java]: I had to modify the original change as we need to support `invocationCount` in `IllegalMonitorTest`.

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java [hunk 5]
```diff
@@ -33,12 +26,15 @@
...
     * keep lists of all the results
     */
-   private AtomicInteger m_numFailed = new AtomicInteger(0);
-   private AtomicInteger m_numSkipped = new AtomicInteger(0);
-   private Map<String, List<ITestResult>> m_allTests = Collections.synchronizedMap(new TreeMap<>());
+   private static final AtomicInteger m_numFailed = new AtomicInteger(0);
+   private static final AtomicInteger m_numSkipped = new AtomicInteger(0);
+   private static final Map<String, PolarionJUnitTest> m_allTests = Collections.synchronizedMap(new TreeMap<>());
+
+   private final int rerunFailingTestsCount = Integer.parseInt(System.getProperty("rerunFailingTestsCount", "0"));
```
Comment (pruivo) [inline commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java:37]: unused?

**Comment by pruivo** [comment]
merged! thanks @ryanemerson !

**Pull request merged**